disorder-fsm/src/Disorder/FSM.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LambdaCase #-}
+module Disorder.FSM (
+    Transition(MkTransition)
+  , always
+  , runFSM
+  ) where
+
+import           Prelude (Show(..))
+import           Data.Bool
+import           Data.String
+import           Data.Maybe
+import           Data.Function
+import           Data.Monoid
+import           Control.Monad
+
+import           Test.QuickCheck.Gen
+import           Test.QuickCheck.Monadic
+import           Test.QuickCheck.Property
+
+-- | Defines a transition from state to state
+data Transition e s m = MkTransition {
+    -- | Display name (used in case of failure)
+    name :: String
+    -- | Predicate which determines of this transition is applicable for a given state 's'
+  , preCond :: s -> Bool
```

Comment:
As mentioned IRL - this looks it could just be a combinator for a given `transition`